fish-rust/build.rs

```diff
@@ -103,12 +113,49 @@ fn detect_features() {
 /// Rust offers fine-grained conditional compilation per-os for the popular operating systems, but
 /// doesn't necessarily include less-popular forks nor does it group them into families more
 /// specific than "windows" vs "unix" so we can conditionally compile code for BSD systems.
-fn detect_bsd() -> miette::Result<bool> {
+fn detect_bsd() -> miette::Result<Option<&'static str>> {
     let uname = std::process::Command::new("uname")
         .output()
         .map_err(|_| miette!("Error executing uname!"))?;
-    Ok(std::str::from_utf8(&uname.stdout)
-        .map(|s| s.to_ascii_lowercase())
-        .map(|s| s.contains("bsd"))
-        .unwrap_or(false))
+    Ok(
+        if std::str::from_utf8(&uname.stdout)
+            .map(|s| s.to_ascii_lowercase())
+            .map(|s| s.contains("bsd"))
+            .unwrap_or(false)
+        {
+            Some("bsd")
+        } else {
+            None
+        },
+    )
+}
+
+fn detect_mb_cur_max() -> miette::Result<Option<&'static str>> {
+    // TODO Respect CXX/CXXFLAGS.
+    let mut child = std::process::Command::new("cc")
+        .args(["-x", "c", "-E", "-pipe", "-"])
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()
+        .unwrap();
+    {
+        let mut stdin = child.stdin.take().unwrap();
+        stdin
+            .write_all(indoc!(
+                b"#include <stdlib.h>
+                  MB_CUR_MAX"
```

Comment:
This breaks cross-compilation, right? Not that I guess it works at the moment anyway.

Reply:
Yeah, it's a shame because Rust makes cross compilation much easier.

We could try to replace `mbrtowc` with Rust's decoding routines (the stdlib supports at least UTF-8 and UTF-16). I'm not sure how complex the requisite locale handling is, and whether we want to support other encodings.

Reply:
I think this only breaks cross-compilation because `cc` is being invoked directly here. We just need to copy-and-paste the code from above to use the `cc-rs` crate to get the compiler and then it should be OK as cross-compilation implies that the headers for the target are automatically called up.

Reply:
thanks for the pointer to `cc-rs`. It looks like we can use it to define a small C library that the rest of the code can link against, so I'll put a portable wrapper around `MB_CUR_MAX` there